### PR TITLE
Add support for period character in table names

### PR DESCRIPTION
**Pull request by villebro**
### CATEGORY

- [x] Bug Fix
- [x] Enhancement (new features, refinement)
- [x] Refactor
- [x] Add tests
- [ ] Build / Development Environment
- [x] Documentation

### SUMMARY
In SQL Lab, table names are currently assumed to follow the following convention:
- `schema.table` or
- `table`

This is handled in the frontend by assuming that a period in the table name always implies a separator between schema and table name. Since there is no standardized way for SQLAlchemy inspectors to return table names (some return `schema.table`, others only `table`), they can be in either format. Since some databases (at least Apache Drill and Postgres) support periods in schema and table names, and their respective inspectors don't return schema prefixed table names, this causes problems when querying tables, as `TableSelector` strips away everything before the period character. This PR moves this logic from the frontend to the backend, and makes it possible to configure this behavior per engine.

This proposal changes table name handling in the following way:
1. An attribute `try_remove_schema_from_table_name` is added to `db_engine_specs` (defaults to `True`). By default, when `True`, `get_table_names()` and `get_view_names()` checks if a table name starts with the schema name followed by a period, and if so, removes the schema name from the table name. Example: `schema.table` becomes `table`, while `table` remains unchanged.
2. Table names are passed as dicts `{'schema': 'schema_name', 'table': 'table_name'}` when handed to the frontend. Previously they were either of the format `table` or `schema.table`. This removes any ambiguity in the frontend.
3. SQL Lab UI now works in the following way:
   - If no schema is selected, table/view names are displayed in the dropdown as `schema.table`.
   - If a schema is selected, tables are shown as `table` only in the dropdown.
Filtering also supports this, i.e. when no schema is chosen, the filter substring makes the comparison assuming that the table name is `schema.table`, making it possible to include the schema name in the filter string.

### SCREENSHOTS
When no schema is chosen, table names are displayed as `schema.table`:
Screenshot 2019-05-09 at 21 09 07

If a schema is selected, only the table name is shown (in this case the table name is `test.table`, not `table` in schema `test`):
Screenshot 2019-05-09 at 21 09 43

### TEST PLAN
Tested locally on Postgres and sqlite. Js unit tests updated to correspond to new data structures and python unit tests added to test table name fetching.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@cgivre

### CODE REVIEW

**Comment by villebro** [comment]
@cgivre please take a look at this PR. If this works out this should make merging your Drill PR slightly easier.

**Comment by cgivre** [comment]
@villebro I'll take a look this weekend.  This pretty much will solve the issues with the Drill integration.

File: superset/db_engine_specs.py
Comment (villebro) [inline superset/db_engine_specs.py]: The reason `db` wasn't type annotated is because it caused a circular import (`models.core` already has a reference to `db_engine_specs`). This should probably be refactored, but is outside the scope of this PR.

**Comment by cgivre** [comment]
@villebro I tried this out and it worked really well even without the Drill PR!  I think pretty much the only thing that Drill needs now for the integration is the time grains!  Thanks for your help with this.  
LGTM +1

**Comment by villebro** [comment]
Thanks for verifying that this works @cgivre . @mistercrunch @john-bodley @betodealmeida would really appreciate help reviewing this, as I think this is one of the last hurdles to being able to support engines that rely on the presence of non-standard characters in schema/table names.

**Comment by villebro** [comment]
Kind reminder to committers that this is pending review, would be great to get this reviewed/merged so Drill integration can be finalized (is blocked by this PR).

**Comment by john-bodley** [comment]
@villebro my main comment which is somewhat related to https://github.com/apache/incubator-superset/pull/7490 is given that the cluster/schema/table name construct can be quite complicated and historically we've often flatten these names into a single string (and then split or used regular expressions to extract the components) whether we should move towards using a class (possibly a `dataclass`) to represent these objects in a canonical way.

**Comment by villebro** [comment]
Good point @john-bodley , having a dedicated class with proper parsing/formatting functionality would probably be a good idea. Do you feel this should be addressed as part of this PR, or start a new PR for that?

**Comment by john-bodley** [comment]
I think a separate PR is fine as it’s probably a large change.

**Review by john-bodley — APPROVED** [review]

**Pull request merged**
